File: sample-python-code/linked_list.py
## debugging a linkedlist structure
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
    
class LinkedList:
    def __init__(self):
        self.root = Node(0)
        self.head = self.root
        self.tail = self.root
    
    def add(self, val):
        new_node = Node(val)
        self.tail.next = new_node
        self.tail = new_node
    
    
    def remove(self, val):
        current = self.head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
                if current.next is None:  # Removed tail
                    self.tail = current
                return True
            current = current.next
        return False
    
if __name__ == "__main__":
    ll = LinkedList()
    
    print("Creating a new linked list...")
    print("\nAdding elements: 10, 20, 30, 40, 50")
    ll.add(10)
    ll.add(20)
    ll.add(30)
    ll.add(40)
    breakpoint()
    ll.add(50)
    
            

        



            

File: sample-cpp-code/linked_list.cpp
#include <iostream>
#include <string>

struct Node {
    int value;
    Node* next;
    
    Node(int val) : value(val), next(nullptr) {}
};

class LinkedList {
public:
    Node* head;
    Node* tail;
    
    LinkedList() : head(nullptr), tail(nullptr) {}
    
    void append(int value) {
        Node* newNode = new Node(value);
        
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }
    
    void remove(int value) {
        if (!head) return;
        
        if (head->value == value) {
            Node* temp = head;
            head = head->next;
            delete temp;
            if (!head) tail = nullptr;
            return;
        }
        
        Node* current = head;
        while (current->next) {
            if (current->next->value == value) {
                Node* temp = current->next;
                current->next = current->next->next;
                if (!current->next) tail = current;
                delete temp;
                return;
            }
            current = current->next;
        }
    }
    
    void print() {
        Node* current = head;
        while (current) {
            std::cout << current->value << " -> ";
            current = current->next;
        }
        std::cout << "null" << std::endl;
    }
    
    ~LinkedList() {
        Node* current = head;
        while (current) {
            Node* temp = current;
            current = current->next;
            delete temp;
        }
    }
};

int main() {
    LinkedList list;
    
    std::cout << "Creating linked list..." << std::endl;
    list.append(10);
    list.append(20);
    list.append(30);
    list.append(40);
    
    std::cout << "List contents: ";
    list.print();
    
    // Set breakpoint here to visualize the structure
    int breakpoint_here = 0;
    
    list.append(50);
    list.print();
    
    return 0;
}

